store/db.go

```diff
@@ -29,9 +30,42 @@ const (
 	DbFilename = "ql.db"
 )
 
+// dbLock is used to guarantee both thread-safety and process-safety
+// for db access.
+type dbLock struct {
+	fl *lock.FileLock
+	// Although the FileLock already ensures thread safety, the Go runtime is unaware
+	// of this, and so Mutex is necessary to satisfy the Go race detector
+	sync.Mutex
+}
+
+func newDBLock(dirPath string) (*dbLock, error) {
+	l, err := lock.NewLock(dirPath, lock.Dir)
+	if err != nil {
+		return nil, err
+	}
+	return &dbLock{fl: l}, nil
+}
+
+func (dl *dbLock) lock() error {
+	if err := dl.fl.ExclusiveLock(); err != nil {
+		return err
+	}
+	dl.Lock()
+	return nil
+}
+
+func (dl *dbLock) unlock() error {
+	if err := dl.fl.Unlock(); err != nil {
+		return err
+	}
+	dl.Unlock()
```

Comment:
@krnowak Why we want to unlock the mutex when we failed to unlock flock?
https://github.com/coreos/rkt/pull/1892#discussion_r48092439

I think from man flock(2), unlocking a flock could return EBADF EINTR EINVAL. If we use the dbLock correct here, I can't imagine there should be failures except for memory flips. So I think when a failure happens, we should be able to retry. However if we move the mutex unlock above, we can't retry.